polytope_server/common/observability/otel.py

```diff
@@ -0,0 +1,86 @@
+from opentelemetry import trace
+from opentelemetry.propagate import inject, extract
+from contextlib import contextmanager
+import logging
+
+def add_trace_context(request):
+    carrier = {}
+
+    # Inject the context into the carrier
+    inject(carrier)
+
+    # Adding trace_id and span_id to the request data for logging
+    request.otel_trace_ctx = {
+        'carrier': carrier  # Store the injected carrier
+    }
+
+    current_span = trace.get_current_span()
+    current_span.set_attribute("polytope.request.id", request.id)
+
+    logging.debug(f"[OTEL] Request created with carrier: {request.otel_trace_ctx['carrier']}")
+
+def update_trace_context(request):
+    carrier = {}
+
+    # Inject the context into the carrier
+    inject(carrier)
+
+    # Adding trace_id and span_id to the request data for logging
+    request.otel_trace_ctx = {
+        'carrier': carrier  # Store the injected carrier
+    }
+
+    logging.debug(f"[OTEL] Request created with carrier: {request.otel_trace_ctx['carrier']}")
+
+
+def restore_trace_context(request):
+    # If otel context is not set, return
+    if not hasattr(request, 'otel_trace_ctx'):
+        return
+
+    logging.debug(f"[OTEL] Restoring context from carrier: {request.otel_trace_ctx['carrier']}")
+
+    # Extract the context from the stored request's carrier
+    extracted_context = extract(carrier=request.otel_trace_ctx['carrier'])
+
+    current_span = trace.get_current_span()
+    current_span.set_attribute("polytope.request.id", request.id)
+
+    return extracted_context
+
+@contextmanager
+def create_new_span_internal(span_name, request_id=None, parent_context=None, kind=trace.SpanKind.INTERNAL):
```

Comment:
If keeping this as four methods, make it clear that `_internal` is referring to the span and not that this is a way of marking this as an internal method (or perhaps someone more experience with Python would understand this immediately?)

Reply:
That's a great question. From my perspective, when discussing OpenTelemetry and spans, it should be clear that we're referring to the type of the span, as that's its primary characteristic. 

Given the context of OpenTelemetry, I believe it would be intuitive for someone familiar with the framework to understand that `_internal` refers to the span type rather than indicating a private or internal method.